src/test/rgw/test_d4n_filter.cc

```diff
@@ -0,0 +1,2810 @@
+      EXPECT_EQ(driver->load_bucket(env->dpp, b, &testBucket, optional_yield{yield}), -2);
+      ASSERT_EQ(testBucket->create(env->dpp, createParams, optional_yield{yield}), 0);
+      testBucket->get_info().bucket.bucket_id = "test_bucket_" + name;
+    }
+
+    void put_object(std::string name, boost::asio::yield_context yield) {
+      std::string object_name = "test_object_" + name;
+      obj = testBucket->get_object(rgw_obj_key(object_name));
+      ASSERT_NE(obj.get(), nullptr);
+      obj.get()->set_obj_size(9);
+
+      testWriter = driver->get_atomic_writer(env->dpp, 
+					      optional_yield{yield},
+					      obj.get(),
+					      acl_owner,
+					      nullptr,
+					      0,
+					      "test_filter");
+
+      const req_context rctx{env->dpp, optional_yield{yield}, nullptr};
+      ceph::real_time mtime; 
+
+      buffer::list bl;
+      bl.append("test_version");
+      rgw::sal::Attrs attrs{{RGW_ATTR_ID_TAG, bl}};
+      bl.clear();
+      bl.append("test_etag");
+      attrs.insert({RGW_ATTR_ETAG, bl});
+
+      bl.clear();
+      bl.append("test data");
+      ASSERT_EQ(testWriter->prepare(optional_yield{yield}), 0);
+      ASSERT_EQ(testWriter->process(std::move(bl), 0), 0);
+      ASSERT_EQ(testWriter->complete(ofs, etag,
+				     &mtime, real_time(),
+				     attrs, std::nullopt,
+				     real_time(),
+				     nullptr, nullptr, nullptr,
+				     nullptr, nullptr, rctx, 0), 0);
+    }
+
+    void put_version_enabled_object(std::string name, std::string& instance, boost::asio::yield_context yield) {
+      testBucket->get_info().flags |= BUCKET_VERSIONED;
+      std::string object_name = "test_object_" + name;
+      objEnabled = testBucket->get_object(rgw_obj_key(object_name));
+      ASSERT_NE(objEnabled.get(), nullptr);
+      objEnabled.get()->set_obj_size(9);
+      objEnabled->gen_rand_obj_instance_name();
+      instance = objEnabled->get_instance();
+
+      testWriter = driver->get_atomic_writer(env->dpp, 
+					      optional_yield{yield},
+					      objEnabled.get(),
+					      acl_owner,
+					      nullptr,
+					      0,
+					      "test_filter");
+
+      const req_context rctx{env->dpp, optional_yield{yield}, nullptr};
+      ceph::real_time mtime; 
+
+      buffer::list bl;
+      bl.append("test_version");
+      rgw::sal::Attrs attrs{{RGW_ATTR_ID_TAG, bl}};
+      bl.clear();
+      bl.append("test_etag");
+      attrs.insert({RGW_ATTR_ETAG, bl});
+
+      bl.clear();
+      bl.append("test data");
+      ASSERT_EQ(testWriter->prepare(optional_yield{yield}), 0);
+      ASSERT_EQ(testWriter->process(std::move(bl), 0), 0);
+      ASSERT_EQ(testWriter->complete(ofs, etag,
+				     &mtime, real_time(),
+				     attrs, std::nullopt,
+				     real_time(),
+				     nullptr, nullptr, nullptr,
+				     nullptr, nullptr, rctx, 0), 0);
+    }
+
+    void put_version_suspended_object(std::string name, boost::asio::yield_context yield) {
+      testBucket->get_info().flags |= BUCKET_VERSIONS_SUSPENDED;
```

Comment:
Is there no method to enable/suspend versioning (instead of directly accessing the flags?)

Reply:
This is how versioning is modified in the posix test. I asked Dan about how to change bucket versioning but this is the only way he knew how. I couldn't find other examples, either

Reply:
Ok